### ISSUE TITLE
Define an IPFS Core spec

### ISSUE THREAD

**Issue by noffle** [ipfs/specs#85]
### Goal

Having a well-defined specification for the IPFS Core API will let us build JS, Go, etc. modules that define the core interface, allowing many different implementations. This is very important for IPFS App developers: rather than hard-code for an HTTP daemon or a Go ephemeral node or a localhost daemon, they can simply develop against the Core API and swap out IPFS Nodes as needed. This will be especially useful once js-ipfs runs in the browser and apps will want to swap between HTTP and a local node.

### Steps

- [ ] update [the core spec](https://github.com/ipfs/specs/tree/1a17d32426f907fb125795e88aca5fa6003f642c/api/core) with API details (use go-ipfs as a base)
- [ ] write a JS ipfs-core-api module that exposes the core interface (with tests)
- [ ] write a Go ipfs-core-api package that exposes the core interface (with tests)
- [ ] update [js-ipfs core](https://github.com/ipfs/js-ipfs/blob/master/src/core/index.js) to adhere to ipfs-core-api
- [ ] update [js-ipfs-api](https://github.com/ipfs/js-ipfs-api) asking that it adhere to the ipfs-core-api

**Comment by diasdavid** [comment]
In addition to the list advantages, having this kind of spec will create good documentation of the several subsystems present inside IPFS, by documenting their entry points.

Agreed with all the steps listed :)

**Comment by RichardLitt** [comment]
Also agreed. What sort of API details do you mean for the core spec?

**Comment by noffle** [comment]
I'm glad you asked. :grinning: 

Right now the [core spec doc](https://github.com/ipfs/specs/tree/1a17d32426f907fb125795e88aca5fa6003f642c/api/core) only lists method names. We should fill in the API method signatures and their semantics. You've done similar work in solidifying the HTTP API, right @RichardLitt?

**Comment by RichardLitt** [comment]
I agree entirely. Give me a syntax and I'll give you a stab. :hocho:

**Comment by noffle** [comment]
I don't know of any language-agnostic API specification formats, so I'd personally be inclined to write [something Node-like](https://www.npmjs.com/package/abstract-blob-store#api) for v1, but generalize out the language-specific bits (like callbacks vs sync methods, `opts` objects, etc).

**Comment by dignifiedquire** [comment]
I think it should be based on a language with a good type system, haskell and rust come to mind, go types would work as well or if you want to keep it javascripty you could use typescript or flowtype.

**Comment by noffle** [comment]
To avoid bikeshed, whoever writes the doc should probably just use the format that feels most comfortable to them. :grinning:

**Comment by lamarpavel** [comment]
As an example, the API spec of OpenGL (and now also Vulkan) are using C syntax as it is common to most programmers and can be seen as somewhat neutral ground. Making it understandable to most programmers could be more important than making it look elegant.

**Comment by diasdavid** [comment]
++ @jbenet @whyrusleeping 

https://github.com/ipfs/specs/tree/master/api/core

**Comment by noffle** [comment]
Good reference from @dignifiedquire: https://github.com/ipfs/js-ipfs-api/blob/feat/files-api/API.md

**Comment by noffle** [comment]
@jbenet @diasdavid @whyrusleeping (and anyone else interested): let's set a time in the near future to do a hangout on the core spec (this issue).

Could you share your availabilities here? I know the next week is going to be busy, but maybe we could do it in person in Berlin and also stream it as a hangout to the rest of the community. Please share your thoughts here.

**Comment by RichardLitt** [comment]
I'm free next week. I want to be there for this.

**Comment by lgierth** [comment]
Core API and go-ipfs-core-api: ipfs/go-ipfs#2876
